src/client/render/secondstage.cpp

```diff
@@ -88,31 +88,70 @@ void PostProcessingStep::run(PipelineContext &context)
 	driver->drawVertexPrimitiveList(&vertices, 4, &indices, 2);
 }
 
+void PostProcessingStep::setBilinearFilter(u8 index, bool value)
+{
+	assert(index < video::MATERIAL_MAX_TEXTURES);
+	material.TextureLayer[index].BilinearFilter = value;
+}
+
 RenderStep *addPostProcessing(RenderPipeline *pipeline, RenderStep *previousStep, v2f scale, Client *client)
 {
 	auto buffer = pipeline->createOwned<TextureBuffer>();
-	static const u8 TEXTURE_COLOR = 0;
-	static const u8 TEXTURE_DEPTH = 3;
+	auto driver = client->getSceneManager()->getVideoDriver();
 
-	// init post-processing buffer
-	buffer->setTexture(TEXTURE_COLOR, scale, "3d_render", video::ECF_A8R8G8B8);
+	// configure texture formats
+	video::ECOLOR_FORMAT color_format = video::ECF_A8R8G8B8;
+	if (driver->queryTextureFormat(video::ECF_A16B16G16R16F))
+		color_format = video::ECF_A16B16G16R16F;
 
 	video::ECOLOR_FORMAT depth_format = video::ECF_D16; // fallback depth format
-	auto driver = client->getSceneManager()->getVideoDriver();
 	if (driver->queryTextureFormat(video::ECF_D32))
 		depth_format = video::ECF_D32;
 	else if (driver->queryTextureFormat(video::ECF_D24S8))
 		depth_format = video::ECF_D24S8;
-	buffer->setDepthTexture(TEXTURE_DEPTH, scale, "3d_depthmap", depth_format);
+
+
+	// init post-processing buffer
+	static const u8 TEXTURE_COLOR = 0;
+	static const u8 TEXTURE_DEPTH = 1;
+	static const u8 TEXTURE_BLOOM = 2;
+	static const u8 TEXTURE_BLUR = 3;
+
+	buffer->setTexture(TEXTURE_COLOR, scale, "3d_render", color_format);
+	buffer->setTexture(TEXTURE_DEPTH, scale, "3d_depthmap", depth_format);
 
 	// attach buffer to the previous step
-	previousStep->setRenderTarget(buffer);
+	previousStep->setRenderTarget(pipeline->createOwned<TextureBufferOutput>(buffer, std::vector<u8> { TEXTURE_COLOR }, TEXTURE_DEPTH));
 
 	// post-processing stage
-	// set up shader
-	u32 shader_id = client->getShaderSource()->getShader("second_stage", TILE_MATERIAL_PLAIN, NDT_MESH);
+	// set up bloom
+	if (g_settings->getBool("enable_bloom"))
+	{
+		buffer->setTexture(TEXTURE_BLUR, scale * 0.5, "blur", color_format);
```

Comment:
I find this downscaling quite reduces the quality of the "god rays".
Can be compensated by the bloom radius, but it's not quite the same, they are naturally more coarse.

Reply:
NM... This does not include "god rays".
IMHO, that could also be done in single PR.

Reply:
Rays need more work and might be done in a pass of their own.